Add metrics documention (#2232)
Azure/azure-service-operator: docs/hugo/content/introduction/metrics.md

# ASOv2 Metrics

Prometheus metrics are exposed for ASOv2 which can be helpful in diagnosability and tracing.
The metrics exposed fall into two groups: Azure based metrics and reconciler metrics (from `controller-runtime`) 

> combined.

Reviewer: I think drop this "Combined"?

## Enabling the metrics

> By default, metrics scraping is turned on for ASOv2, which can be turned off using helm chart or Deployment yaml.

Reviewer: ```suggestion
By default, metrics scraping is turned on for ASOv2. It can be turned off using the Helm chart <What parameter?> or by editing the Deployment yaml <what parameter>?
```

### ASOv2 Helm Chart

> While installing the Helm chart, we can turn the metrics ON and OFF and set the metrics expose address using the 

Reviewer 2: Suggest using markdown for emphasis instead of capitals.

```suggestion
While installing the Helm chart, we can turn the metrics _**on**_ and _**off**_ and set the metrics expose address using the 
```

below settings. Also, we can change the settings inside `values.yaml` file for ASOv2 Helm chart.

   ```
   --set metrics.enable=true/false (default: true)
   --set metrics.address=127.0.0.1:8080 (default)

>    ```

Reviewer 2: Should this be indented?

### Deployment Yaml

> In the deployment yaml, we can turn OFF the metrics by omitting the `metrics-addr` flag. We can also change to use 

Reviewer 2: ```suggestion
In the deployment yaml, we can turn _**off**_ the metrics by omitting the `metrics-addr` flag. We can also change to use 
```

> In the deployment yaml, we can turn OFF the metrics by omitting the `metrics-addr` flag below and also set the metrics 

Reviewer: minor: fix this sentence. You sorta same the same thing twice. I think you can just remove the "and also ..." part.

expose address there.
    
   ```
    spec:
      containers:
      - args:
        - --metrics-addr=127.0.0.1:8080 (default)
   ```
## Understanding the ASOv2 Metrics

> | Metric                                         | Description                                                                                               | Label 1      | Label 2    | Label 3      |

Reviewer: Nice, this table is super useful I think

|------------------------------------------------|--------------------------------------------------------------------------------------------------------------|--------------|-------------|--------------|

> | `controller_runtime_reconcile_total`           | A prometheus counter metric with total number of reconcilations per controller.                              | Controller   | Result      |              |

Reviewer 2: Is this metric generated independently for each controller, or just once for all controllers?

Might also need to define what "controller" means - just something that says "Each resource being reconciled against Azure ARM has a separate dedicated controller."

Author: Done! Added another section which explains all the labels we're using

| `controller_runtime_errors_total`              | A prometheus counter metric with total number of errors from reconciler                                      | Controller   |             |              |
| `controller_runtime_reconcile_time_seconds`    | A prometheus histogram metric which keeps track of the duration of reconcilations                            | Controller   |             |              |
| `controller_runtime_max_concurrent_reconciles` | A prometheus gauge metric with the number of concurrent reconciles per controller                            | Controller   |             |              |
| `controller_runtime_active_workers`            | A prometheus gauge metric with the number of active workers per controller                                   | Controller   |             |              |

> | `azure_requests_total`                         | A prometheus counter metric with total number of requests to Azure                                           | ResourceName | RequestType | ResponseCode |

Reviewer: minor: "with the", rather than just "with"? Same on the line below.

| `azure_failed_requests_total`                  | A prometheus counter metric with the record of the number of failed requests to ARM.                      | ResourceName | HTTPMethod |              |

> | `azure_requests_time_seconds`                  | A prometheus histogram metric with the record the round-trip time taken by the request to ARM             | ResourceName | HTTPMethod |              |

Reviewer: ```suggestion
| `azure_requests_time_seconds`                  | A prometheus histogram metric with the round-trip time taken by the request to ARM             | ResourceName | HTTPMethod |              |
```

# ASOv2 Metrics

Prometheus metrics are exposed for ASOv2 which can be helpful in diagnosability and tracing.

> The metrics exposed fall into two groups: Azure based metrics and reconciler metrics (from `controller-runtime`).

Reviewer 2: Good place to use an Oxford comma for clarity.

```suggestion
The metrics exposed fall into two groups: Azure based metrics, and reconciler metrics (from `controller-runtime`).
```

## Disabling the metrics

By default, metrics scraping is turned on for ASOv2. It can be turned off using the Helm chart 

> `--set metrics.enable` or by omitting `--metrics-addr` flag from the Deployment yaml.

Reviewer 2: Why does `--set metrics.enable` turn OFF metrics scraping?

Reviewer 2: Oh, you cover the details below. Perhaps reword this paragraph to make it read more like an introduction.

Author: Makes sense.. I've made it more clear in the recent commit.
